queries/stack-graphs.tsg

```diff
@@ -0,0 +1,1756 @@
+;; unqualified references".
+[
+  ;; ### Types
+  (abstract_type trait: (_) @here)
+  (array_type element: (_) @here)
+  (bounded_type (_) @here)
+  (dynamic_type trait: (_) @here)
+  (function_type trait: (_) @here)  ;; only for Fn, FnMut, FnOnce
+  (function_type return_type: (_) @here)
+  (generic_type type: (_) @here)
+  (generic_type_with_turbofish type: (_) @here)
+  (pointer_type type: (_) @here)
+  (qualified_type type: (_) @here)
+  (qualified_type alias: (_) @here)
+  (reference_type type: (_) @here)
+  (scoped_type_identifier path: (_) @here)
+  (tuple_type (_) @here)
+  (type_binding type: (_) @here)
+
+  ;; ### Expressions
+  (type_cast_expression type: (_) @here)
+  (bracketed_type (_) @here)
+  (struct_expression name: (_) @here)
+  (scoped_identifier path: (_) @here)
+
+  ;; ### Patterns
+  (struct_pattern type: (_) @here)
+
+  ;; ### Items
+  (const_item type: (_) @here)
+  (field_declaration type: (_) @here)
+  (function_item return_type: (_) @here)
+  (impl_item trait: (_) @here)
+  (impl_item type: (_) @here)
+  (ordered_field_declaration_list type: (_) @here)
+  (parameter type: (_) @here)
+  (static_item type: (_) @here)
+  (trait_bounds (_) @here)
+  (type_item type: (_) @here)
+  (where_predicate left: (_) @here)
+
+  ;; ### Statements
+  (let_declaration type: (_) @here)
+] {
+  let @here.context = "%types"
+}
+
+;; Ambiguous contexts
+(type_arguments (_) @here) {
+  ;; type or value depending on the generic, I think
+  let @here.context = "%types" ;; bug
+}
+
+(tuple_struct_pattern type: (_) @here) {
+  ;; type or value, whichever comes first
+  let @here.context = "%values" ;; BUG - try "%any"
+}
+
+;; Pattern context
+[
+  (for_expression pattern: (_) @here)
+  (if_let_expression pattern: (_) @here)
+  (while_let_expression pattern: (_) @here)
+  (let_declaration pattern: (_) @here)
+  (parameter pattern: (_) @here)
+  (closure_parameters (_) @here)
+  (match_pattern . (_) @here)
+  (tuple_pattern (_) @here)
+  (tuple_struct_pattern type: (_) (_) @here)
+  (slice_pattern (_) @here)
+  (field_pattern name: (shorthand_field_identifier) @here)
+  (field_pattern pattern: (_) @here)
+  (mut_pattern (mutable_specifier) (_) @here)
+  (ref_pattern (_) @here)
+  (captured_pattern (_) @here "@" (_))
+  (reference_pattern "&" (mutable_specifier)? (_) @here)
+  (or_pattern (_) @here)
+] {
+  let @here.context = "%pattern"
+}
+
+;; Macro context
+(macro_invocation macro: (_) @here) {
+  let @here.context = "%macro"
+}
+
+;; Use-path context
+[
+  (scoped_use_list path: (_) @here)
+  (use_as_clause path: (_) @here)
+  (use_declaration argument: (_) @here)
+  (use_list (_) @here)
+  (use_wildcard (_) @here "::" "*")
+] {
+  ;; Here we may be importing either a type or a value, or both. We can't
+  ;; locally tell which, so importable scopes must provide that `%any` reaches
+  ;; both the `%values` namespace and the `%types` namespace.
+  let @here.context = "%any"
```

Comment:
I am not sure the use of an `%any` namespace is a good idea here. If I understand it correctly, this is not an "actual" namespace in Rust, but a union of `%types` and `%values`.

An alternative approach could be to introduce two separate paths, one pushing `%types` and another `%values`. Depending on how this is used, perhaps precedence values are enough to ensure the correct one is chosen?

Maybe this is still not enough, but I expect it is equivalent to making the union explicit everywhere.

Reply:
You're right that `%any` is
- a union of `%types` and `%values`;
- a not-very-good hack.

I would love to implement your suggestion, I just couldn't figure out how.

(However, note that precedence is unwanted here--if I write `use foo::X;` and there is both a type `X` and a separately declared value `X` in the module `foo`, then it is correct to offer both definitions.)

Reply:
OK, I think I may have figured out an approach here, but let's fix it in a later PR.